In [41]:
import pandas as pd
import numpy as np
import random
from datetime import datetime, timedelta

np.random.seed(42)
random.seed(42)

start_id = 21
num_rows = 600

departments = ["IT", "Sales", "Finance", "HR", "Operations"]

roles = {
    "IT": ["Software Engineer", "Data Analyst", "Data Engineer", "Business Analyst", "Software Developer"],
    "Sales": ["Sales Executive", "Sales Manager", "Sales Associate"],
    "Finance": ["Accountant", "Financial Analyst", "Auditor", "Treasury Analyst"],
    "HR": ["HR Manager", "HR Executive", "Recruiter", "HR Coordinator"],
    "Operations": ["Team Lead", "Supervisor", "Logistics Manager", "Operations Analyst"]
}

education_map = {
    "IT": ["B.Tech", "B.Sc", "M.Tech"],
    "Sales": ["BBA", "MBA"],
    "Finance": ["B.Com", "MBA"],
    "HR": ["BA", "MBA"],
    "Operations": ["BBA", "MBA"]
}

locations = ["Bangalore", "Mumbai", "Delhi", "Hyderabad", "Pune"]

data = []

for i in range(num_rows):
    emp_id = start_id + i
    gender = random.choice(["Male", "Female"])
    name = f"Employee_{emp_id}"   # simple generated name
    department = random.choice(departments)
    role = random.choice(roles[department])
    salary = random.randint(40000, 100000)
    
    hire_date = datetime.now() - timedelta(days=random.randint(0, 3650))
    age = random.randint(23, 45)
    education = random.choice(education_map[department])
    performance_score = round(random.uniform(3.0, 5.0), 1)
    attrition = random.choice(["Yes", "No"])
    location = random.choice(locations)

    exit_date = None
    if attrition == "Yes":
        days_worked = (datetime.now() - hire_date).days
        if days_worked > 30:
           exit_date = hire_date + timedelta(days=random.randint(30, days_worked))
           exit_date = exit_date.date()
    else:
        attrition = "No" 

    data.append([
        emp_id, name, gender, department, role, salary,
        hire_date.date(), exit_date, age, education,
        performance_score, attrition, location
    ])

columns = [
    "employee_id", "name", "gender", "department", "role",
    "salary", "hire_date", "exit_date", "age",
    "education", "performance_score", "attrition", "location"
]

df_new = pd.DataFrame(data, columns=columns)

df_new.to_csv("employee_data_600_rows.csv", index=False)

print("600 rows generated successfully.")

600 rows generated successfully.


In [42]:
df_new

,employee_id,name,gender,department,role,salary,hire_date,exit_date,age,education,performance_score,attrition,location
0,21,Employee_21,Male,IT,Data Engineer,56049,2023-08-25,2024-11-29,27,M.Tech,3.2,Yes,Pune
1,22,Employee_22,Male,IT,Software Engineer,54328,2023-07-18,2024-10-19,39,M.Tech,3.1,Yes,Pune
2,23,Employee_23,Male,HR,Recruiter,93046,2016-05-27,None,23,BA,4.4,No,Delhi
3,24,Employee_24,Male,Sales,Sales Manager,46698,2025-02-10,None,35,BBA,3.7,No,Pune
4,25,Employee_25,Female,IT,Business Analyst,75142,2024-10-01,None,35,B.Tech,4.1,No,Pune
...,...,...,...,...,...,...,...,...,...,...,...,...,...
595,616,Employee_616,Female,HR,HR Manager,40269,2022-08-31,2025-08-31,24,BA,3.7,Yes,Bangalore
596,617,Employee_617,Female,IT,Software Developer,89708,2022-08-26,2024-04-13,43,B.Tech,3.6,Yes,Delhi
597,618,Employee_618,Female,HR,HR Manager,57868,2024-06-23,2026-01-29,28,MBA,3.2,Yes,Mumbai
598,619,Employee_619,Female,Operations,Logistics Manager,57763,2024-07-25,2025-07-08,45,MBA,4.1,Yes,Bangalore


In [53]:
df_new.to_csv(r'C:\Users\LENOVO\Desktop\HR analysis\hr_analytics_new.csv',index=True)

In [52]:
df_new

,employee_id,name,gender,department,role,salary,hire_date,exit_date,age,education,performance_score,attrition,location
0,21,Employee_21,Male,IT,Data Engineer,56049,2023-08-25,2024-11-29,27,M.Tech,3.2,Yes,Pune
1,22,Employee_22,Male,IT,Software Engineer,54328,2023-07-18,2024-10-19,39,M.Tech,3.1,Yes,Pune
2,23,Employee_23,Male,HR,Recruiter,93046,2016-05-27,None,23,BA,4.4,No,Delhi
3,24,Employee_24,Male,Sales,Sales Manager,46698,2025-02-10,None,35,BBA,3.7,No,Pune
4,25,Employee_25,Female,IT,Business Analyst,75142,2024-10-01,None,35,B.Tech,4.1,No,Pune
...,...,...,...,...,...,...,...,...,...,...,...,...,...
595,616,Employee_616,Female,HR,HR Manager,40269,2022-08-31,2025-08-31,24,BA,3.7,Yes,Bangalore
596,617,Employee_617,Female,IT,Software Developer,89708,2022-08-26,2024-04-13,43,B.Tech,3.6,Yes,Delhi
597,618,Employee_618,Female,HR,HR Manager,57868,2024-06-23,2026-01-29,28,MBA,3.2,Yes,Mumbai
598,619,Employee_619,Female,Operations,Logistics Manager,57763,2024-07-25,2025-07-08,45,MBA,4.1,Yes,Bangalore


In [54]:
df_new['attrition_binary']= df_new['attrition'].map({'Yes':1,'No':0})
df_new.head()

,employee_id,name,gender,department,role,salary,hire_date,exit_date,age,education,performance_score,attrition,location,attrition_binary
0,21,Employee_21,Male,IT,Data Engineer,56049,2023-08-25,2024-11-29,27,M.Tech,3.2,Yes,Pune,1
1,22,Employee_22,Male,IT,Software Engineer,54328,2023-07-18,2024-10-19,39,M.Tech,3.1,Yes,Pune,1
2,23,Employee_23,Male,HR,Recruiter,93046,2016-05-27,None,23,BA,4.4,No,Delhi,0
3,24,Employee_24,Male,Sales,Sales Manager,46698,2025-02-10,None,35,BBA,3.7,No,Pune,0
4,25,Employee_25,Female,IT,Business Analyst,75142,2024-10-01,None,35,B.Tech,4.1,No,Pune,0


In [57]:
baseline_attrition=df_new['attrition_binary'].mean()
baseline_attrition

np.float64(0.5016666666666667)

In [61]:
df_new.groupby('location')['attrition_binary'].mean().sort_values(ascending=True) # banglore
df_new.groupby('education')['attrition_binary'].mean().sort_values(ascending=True) # MBA
df_new.groupby('age')['attrition_binary'].mean().sort_values(ascending=True) # 39
df_new.groupby('gender')['attrition_binary'].mean().sort_values(ascending=True) # female

gender
Male      0.496622
Female    0.506579
Name: attrition_binary, dtype: float64

In [91]:
risky_stuff=((df_new['location']=='Bangalore')&(df_new['education']=='MBA'))
risky=df_new[risky_stuff]
risky.head()
others=df_new[~risky_stuff]
others

,employee_id,name,gender,department,role,salary,hire_date,exit_date,age,education,performance_score,attrition,location,attrition_binary
0,21,Employee_21,Male,IT,Data Engineer,56049,2023-08-25,2024-11-29,27,M.Tech,3.2,Yes,Pune,1
1,22,Employee_22,Male,IT,Software Engineer,54328,2023-07-18,2024-10-19,39,M.Tech,3.1,Yes,Pune,1
2,23,Employee_23,Male,HR,Recruiter,93046,2016-05-27,None,23,BA,4.4,No,Delhi,0
3,24,Employee_24,Male,Sales,Sales Manager,46698,2025-02-10,None,35,BBA,3.7,No,Pune,0
4,25,Employee_25,Female,IT,Business Analyst,75142,2024-10-01,None,35,B.Tech,4.1,No,Pune,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
593,614,Employee_614,Male,Sales,Sales Executive,69083,2020-06-23,2022-06-19,23,MBA,4.8,Yes,Mumbai,1
594,615,Employee_615,Female,Operations,Logistics Manager,90825,2022-06-25,2022-10-07,27,BBA,4.6,Yes,Mumbai,1
595,616,Employee_616,Female,HR,HR Manager,40269,2022-08-31,2025-08-31,24,BA,3.7,Yes,Bangalore,1
596,617,Employee_617,Female,IT,Software Developer,89708,2022-08-26,2024-04-13,43,B.Tech,3.6,Yes,Delhi,1


In [100]:
risk_rate=risky['attrition_binary'].mean()

In [99]:
others_rate=others['attrition_binary'].mean()

In [102]:
risk_multiplier=risk_rate/others_rate
risk_multiplier

np.float64(1.3096917450365728)

In [103]:
round(risk_multiplier,2)

np.float64(1.31)

In [116]:
risky_rate=round(risk_rate*100,2)
a=pd.DataFrame([risky_rate],columns=['risky_rate'])
a.to_csv(r'C:\Users\LENOVO\Desktop\HR analysis\risky_rate.csv',index=True)

In [123]:
other_rate=round(others_rate*100,2)
b=pd.DataFrame([other_rate],columns=['other_rate'])
b.to_csv(r'C:\Users\LENOVO\Desktop\HR analysis\other_rate.csv',index=True)

In [130]:
c=pd.concat([a,b],axis=1)
c.to_csv(r'C:\Users\LENOVO\Desktop\HR analysis\risk_rate.csv',index=True)

In [131]:
c

,risky_rate,other_rate
0,63.79,48.71


**Employees who are 39 and Female with BBA in Bangalore have a 63% attrition rate,
compared to 48.71% of other employees making them 1.13% more likely to exit**